### ISSUE TITLE
Add better Type registry support

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#327]
In a recent-ish update to GraphQL PHP the concept of a Type Registry was added where Types will be lazy-loaded only when they're needed.

So, we can have a central registry where Types are registered, but GraphQL will only instantiate them if they're needed. 

You can see docs for GraphQL PHP here: http://webonyx.github.io/graphql-php/type-system/schema/#lazy-loading-of-types

For WPGraphQL, I think we should expose a central Type registry, and WPGraphQL will handle registering the core Types there, but make it extendable so Plugins and themes can register their own Types as well. 

Right now, when calling a Type in the Schema, we do something like: 
```
['myField'] => [
   'description' => __( 'Description of my field', 'wp-graphql' ),
   'type' => \WPGraphQL\Types::string(),
]
```
I see this changing to something like: 

```
['myField'] => [
   'description' => __( 'Description of my field', 'wp-graphql' ),
   'type' => \WPGraphQL\TypeRegistry::get('string'),
]
```

And can actually probably be even further abstracted so we can just define Schema like so: 

```
['myField'] => [
   'description' => __( 'Description of my field', 'wp-graphql' ),
   'type' => 'string',
]
```

And anywhere a field declares a Type, we can resolve by calling `\WPGraphQL\TypeRegistry::get( $field['type'] );` so that it doesn't have to be called in the Schema definition itself. 

Not sure if this will lead to any breaking changes or not. I _think_ we can probably support this change without breaking anything, but not 100% sure yet.

**Comment by renatonascalves** [comment]
I'd say the last option would be the best one for me. I'm all for simplicity.

**Comment by jasonbahl** [comment]
See: #288 and #246 for related info. Closing them and consolidating to this ticket.

**Comment by jasonbahl** [comment]
^ reopened this as this is the Issue we're tracking #288 and #246 on. Accidentally closed this when I intended to close those ones.

**Comment by jasonbahl** [comment]
closed by #558
